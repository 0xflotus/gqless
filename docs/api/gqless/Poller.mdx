---
id: Poller
title: Poller
---

> Subject to [limitations of getAccessor](/docs/api/gqless/getAccessor)

Fetches an accessor at a set interval from the network.

## Example

Fetch `query.me.name` every one second:

```ts
import { Poller } from 'gqless'

const poller = new Poller(query.me.name, 1000)
// Start the poller
poller.toggle(true)
```
