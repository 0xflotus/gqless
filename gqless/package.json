{
  "license": "MIT",
  "name": "gqless",
  "version": "0.0.1-alpha.20",
  "main": "dist/index.js",
  "source": "src/index.ts",
  "module": "dist/gqless.esm.js",
  "homepage": "https://gqless.netlify.com",
  "repository": {
    "type": "git",
    "url": "https://github.com/samdenty/gqless.git",
    "directory": "gqless"
  },
  "typings": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "watch": "tsdx watch",
    "build": "cp ../../README.md .; tsdx build --tsconfig ./tsconfig.build.json",
    "test": "jest"
  },
  "dependencies": {
    "@gqless/utils": "^0.0.1-alpha.20",
    "lodash": "^4.17.15",
    "mix-classes": "^1.1.10"
  },
  "devDependencies": {
    "@babel/core": "^7.6.2",
    "@types/lodash": "^4.14.141",
    "template-literal-table": "^2.0.4"
  },
  "gitHead": "7c27436fcea8e672f07233010401417ea7c59760"
}
