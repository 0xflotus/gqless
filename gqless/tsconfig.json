{
  "extends": "../tsconfig.json",
  "include": ["src", "test"],
  "compilerOptions": {
    // https://github.com/parcel-bundler/parcel/issues/2856
    "downlevelIteration": true
  }
}
