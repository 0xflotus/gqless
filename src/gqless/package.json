{
  "name": "gqless",
  "version": "0.0.1-alpha.3",
  "main": "dist/index.js",
  "umd:main": "dist/gqless.umd.production.js",
  "module": "dist/gqless.es.production.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "watch": "tsdx watch",
    "build": "tsdx build",
    "prepublish": "cp ../../README.md .; yarn build",
    "test": "tsdx test"
  },
  "dependencies": {
    "@gqless/utils": "^0.0.1-alpha.3",
    "fast-deep-equal": "^2.0.1",
    "json-stable-stringify": "^1.0.1",
    "mix-classes": "^1.1.6"
  },
  "devDependencies": {
    "@babel/core": "^7.4.4",
    "@types/json-stable-stringify": "^1.0.32",
    "template-literal-table": "^2.0.2"
  },
  "gitHead": "a6601a3f9703830b097595b2eab2ed357f529feb"
}
