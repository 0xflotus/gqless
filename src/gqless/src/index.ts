export * from './Node'
export * from './Cache'
export * from './Selection'
export * from './ASTBuilder'
export * from './GraphQL'
export * from './Plugin'
export * from './Interceptor'
export * from './Accessor'
export * from './Scheduler'
export * from './Type'
export * from './Variable'
export * from './helpers'
