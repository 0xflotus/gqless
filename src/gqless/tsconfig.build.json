{
  "extends": "../../tsconfig.build.json",
  "include": ["src"],
  "compilerOptions": {
    "rootDir": "./"
  },
  "references": [
    {
      "path": "../../packages/utils/tsconfig.build.json"
    }
  ]
}
