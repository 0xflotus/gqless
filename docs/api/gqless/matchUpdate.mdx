---
id: matchUpdate
title: matchUpdate
---

> Subject to [limitations of getAccessor](/docs/api/gqless/getAccessor)

Updates the value for a given accessor, by performing a pattern match against all other values of same type. See [Matchers](/docs/matchers) for more.

## Example

Set `query.me` to a User in the cache, with a name of `bob`:

```ts
import { matchUpdate } from 'gqless'

const updated = matchUpdate(query.me, { name: 'bob' })

if (!updated) {
  throw new Error('bad match')
}
```
