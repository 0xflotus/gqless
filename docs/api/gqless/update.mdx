---
id: update
title: update
---

> Subject to [limitations of getAccessor](/docs/api/gqless/getAccessor)

Updates the data for a given accessor, performing a merge-update (same as when fetched from the network).

## Example

```ts
import { update } from 'gqless'

// Update scalars
update(query.me.name, 'bob')

// Merge in new data (existing fields left untouched)
update(query, { me: { name: 'bob' } })
```
