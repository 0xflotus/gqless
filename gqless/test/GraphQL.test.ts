import { createGraphQL } from 'testing'
import { GraphQL, Variable } from 'gqless'

jest.useFakeTimers()

const commitFetched = async (graphql: GraphQL) => {
  const { commit } = graphql.scheduler
  jest.runOnlyPendingTimers()
  await commit.onFetched
}

it('fetches values', async () => {
  const graphql = createGraphQL()

  graphql.query.string
  graphql.query.arrayOfString[0]
  graphql.query.arrayOfObjects.forEach(object => {
    object.string
  })

  await commitFetched(graphql)

  expect(graphql.query.string).toMatchSnapshot()
  expect(graphql.query.arrayOfString).toMatchSnapshot()
  expect(graphql.query.arrayOfObjects[0].string).toMatchSnapshot()

  expect(graphql.cache.toJSON()).toMatchSnapshot()

  expect(graphql.scheduler.commit.selections.size).toEqual(0)
})

it(`shares cache for arguments with/without variables`, async () => {
  const graphql = createGraphQL()

  const object = graphql.query.arrayOfObjects({ string: 'test' })[0]
  object.string

  await commitFetched(graphql)

  expect(
    graphql.query.arrayOfObjects({ string: 'test' })[0].string
  ).toMatchSnapshot()

  expect(
    graphql.query.arrayOfObjects({ string: new Variable('test') })[0].string
  ).toMatchSnapshot()

  expect(graphql.cache.toJSON()).toMatchSnapshot()

  expect(graphql.scheduler.commit.selections.size).toEqual(0)
})
