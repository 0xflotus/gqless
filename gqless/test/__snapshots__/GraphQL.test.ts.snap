// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches values 1`] = `"Maurine"`;

exports[`fetches values 2`] = `
bound Array [
  "Edison",
  "Hilbert",
  "Danielle",
  "Maryse",
]
`;

exports[`fetches values 3`] = `"Miller"`;

exports[`fetches values 4`] = `
Object {
  "Query": Object {
    "arrayOfObjects": Array [
      Object {
        "string": "Miller",
      },
      Object {
        "string": "Millie",
      },
      Object {
        "string": "Jennings",
      },
    ],
    "arrayOfString": Array [
      "Edison",
      "Hilbert",
      "Danielle",
      "Maryse",
    ],
    "string": "Maurine",
  },
}
`;

exports[`shares cache for arguments with/without variables 1`] = `"Mervin"`;

exports[`shares cache for arguments with/without variables 2`] = `"Mervin"`;

exports[`shares cache for arguments with/without variables 3`] = `
Object {
  "Query": Object {
    "arrayOfObjects({\\"string\\":\\"test\\"})": Array [
      Object {
        "string": "Mervin",
      },
      Object {
        "string": "Edison",
      },
      Object {
        "string": "Hilbert",
      },
      Object {
        "string": "Danielle",
      },
    ],
  },
}
`;
